File: ts/src/abnf.ts
/* Copyright (c) 2025-2026 Richard Rodger and other contributors, MIT License */

/*  abnf.ts
 *  ABNF plugin — adds `tn.abnf(src)` (install) and `tn.abnf.toSpec(src)`
 *  (build without installing) to an Tabnas instance.
 *
 *  The conversion logic itself lives in ./converter.ts; this file
 *  exposes it both as a Plugin (for `tn.use(abnf)`) and as bare
 *  exports (for code that wants to convert without an instance).
 */

import type {
  Tabnas,
  GrammarSpec,
  Plugin,
} from '@tabnas/parser'

import {
  abnf as abnfConvert,
  parseAbnf,
  emitGrammarSpec,
  eliminateLeftRecursion,
  abnfRules,
  AbnfParseError,
  AbnfConvertOptions,
} from './converter'

import {
  abnfCompile,
  toRecognitionSpec,
  toPureSpec,
  toJsonic,
  attachActions,
  attachActionSlots,
  markListing,
  AbnfCompileError,
  AbnfActionError,
  AbnfCompileOptions,
  ActionsMap,
} from './compile'


// Plugin entry point. Decorates the instance with a callable `abnf`
// member that converts and installs a grammar, plus `abnf.toSpec` for
// callers that just want the spec.
type AbnfPluginOptions = AbnfConvertOptions & { actions?: ActionsMap }

const abnf: Plugin = function abnf(tn: Tabnas, _options?: any): void {
  const fn = ((src: string, opts?: AbnfPluginOptions): GrammarSpec => {
    // User actions wrap the compiler's closures, so convert in closure
    // mode (never builtins) and request marks when actions are supplied.
    const spec = abnfConvert(src,
      opts?.actions ? { ...opts, builtins: false, marks: true } : opts)
    if (opts?.actions) attachActions(spec, opts.actions)
    tn.grammar(spec)
    return spec
  }) as ((src: string, opts?: AbnfPluginOptions) => GrammarSpec) & {
    toSpec: (src: string, opts?: AbnfConvertOptions) => GrammarSpec
  }
  fn.toSpec = (src: string, opts?: AbnfConvertOptions): GrammarSpec =>
    abnfConvert(src, opts)
  tn.abnf = fn
}


export {
  abnf,
  abnfConvert,
  parseAbnf,
  emitGrammarSpec,
  eliminateLeftRecursion,
  abnfRules,
  AbnfParseError,
  abnfCompile,
  toRecognitionSpec,
  toPureSpec,
  toJsonic,
  attachActions,
  attachActionSlots,
  markListing,
  AbnfCompileError,
  AbnfActionError,
}

export type { AbnfConvertOptions, AbnfCompileOptions, ActionsMap }

File: go/abnf.go
// Copyright (c) 2025-2026 Richard Rodger and other contributors, MIT License

// Package tabnasabnf is an ABNF -> tabnas GrammarSpec compiler for the
// tabnas parsing engine (github.com/tabnas/parser/go). It is a faithful
// Go port of the @tabnas/abnf TypeScript package.
//
// Given a small ABNF dialect it produces a function-free (when
// requested) GrammarSpec that, installed on a tabnas engine, parses
// inputs in that grammar and builds a {rule, src, kids} AST. It also
// emits "pure-data" jsonic (recognition / pure specs) and supports
// user actions.
//
// The package mirrors the TS sources:
//   - converter.ts -> converter.go (AST, parseAbnf, abnfRules, desugar,
//     core rules, left-recursion elimination, probe-dispatch rewriter,
//     FIRST sets, emitGrammarSpec, token allocation, Abnf entry) and
//     parser_abnf.go (the ABNF-for-ABNF parser grammar).
//   - compile.ts -> compile.go (AbnfCompile, ToRecognitionSpec,
//     ToPureSpec, ToJsonic, AttachActions, MarkListing).
//   - abnf.ts -> facade.go (Abnf, ParseAbnf, EmitGrammarSpec,
//     EliminateLeftRecursion, Install — the public facade).
package tabnasabnf

// VERSION is this module's version. It MUST equal ts/package.json
// "version": the release orchestrator rewrites both, and
// TestVersionMatchesPackageJSON fails the build if they drift.
const VERSION = "0.4.7"
